src/fairseq2/models/llama/attention.py

```diff
@@ -0,0 +1,278 @@
+        v_proj: Optional[Projection] = None,
+        pos_encoder: Optional[PositionEncoder] = None,
+        sdpa: Optional[SDPA] = None,
+        output_proj: Optional[Projection] = None,
+        bias: bool = True,
+        device: Optional[Device] = None,
+        dtype: Optional[DataType] = None,
+    ) -> None:
+        """
+        :param model_dim:
+            The dimensionality of the model.
+        :param num_heads:
+            The number of attention heads.
+        :param num_key_value_heads:
+            The number of key-value heads for Grouped Query Attention.
+            If set to ``num_attn_heads``, it is equivalent to normal
+            Multi Head Attention (MHA). If set to ``1``, it is equivalent to
+            Multi Query Attention (MQA).
+        :param q_proj:
+            The projection to apply to queries before computing attention. If
+            ``None``, a default projection will be used.
+        :param k_proj:
+            The projection to apply to keys before computing attention. If
+            ``None``, a default projection will be used.
+        :param v_proj:
+            The projection to apply to values before computing attention. If
+            ``None``, a default projection will be used.
+        :param pos_encoder:
+            The position encoder to apply to queries and keys after projection.
+        :param sdpa:
+            The scaled dot-product attention module to compute head attentions.
+            If ``None``, a default implementation will be used.
+        :param scale_heads:
+            If ``True``, applies head scaling as described in
+            :cite:t:`https://doi.org/10.48550/arxiv.2110.09456`
+        :param output_proj:
+            The projection to produce final attentions. If ``None``, a
+            default projection will be used.
+        :param bias:
+            If ``True``, query, key, value, and output projections learn an
+            additive bias. Ignored for explicitly specified projections.
+        """
+        super().__init__(model_dim, num_heads)
+
+        if model_dim % num_heads != 0:
+            raise ValueError(
+                f"`model_dim` ({model_dim}) must be divisible by `num_heads` ({num_heads})."
+            )
+
+        if num_heads % num_key_value_heads != 0:
+            raise ValueError(
+                f"`num_heads` ({num_heads}) must be divisible by `num_key_value_heads` ({num_key_value_heads})."
+            )
+
+        self.num_key_value_heads = num_key_value_heads
+        self.head_dim = model_dim // num_heads
+        self.num_key_value_groups = num_heads // num_key_value_heads
+
+        if q_proj is None and k_proj is None and v_proj is None:
+            q_proj = QKVProjection(
+                self.head_dim * num_heads, bias, device=device, dtype=dtype
+            )
+            k_proj = QKVProjection(
+                self.head_dim * num_key_value_heads, bias, device=device, dtype=dtype
```

Comment:
Not entirely sure how this works. Assuming this is part of self attention, the input Q/K/V will all be the same input sequence with a shape of [bsz, seq_len, model_dim]. If that is the case, my intuition says that `k_proj` and `v_proj` should be `Linear(input_dim=model_dim, output_dim=self.head_dim * num_key_value_heads)`.

Reply:
Thank you for catching this!